### ISSUE TITLE
Transfer数据集SUBJ无法测试

### ISSUE THREAD

**Issue by xiami2019** [kongds/Prompt-BERT#23]
您好，很棒的工作！
我在复现时遇到一点问题，再跑SUBJ这个数据集时会出现
/pytorch/aten/src/ATen/native/cuda/IndexKernel.cu:84: operator(): block: [21,0,0], thread: [127,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
这个问题，想请问这个是由什么问题导致的呢，您遇到过类似的问题吗？

**Comment by kongds** [comment]
谢谢，我没有遇到过这个问题
请问可以提供执行时的命令吗？

**Comment by xiami2019** [comment]
直接使用的eval脚本中测试RoBERTa模型的指令进行测试的
`EXP=$1
python evaluation.py \
        --model_name_or_path $EXP\
        --pooler avg\
        --mode test\
        --task_set SUBJ\
        --mask_embedding_sentence \
        --mask_embedding_sentence_use_pooler\
        --mask_embedding_sentence_delta \
        --mask_embedding_sentence_template "*cls*_This_sentence_:_'_*sent_0*_'_means*mask*.*sep+*"`
MPQA这个二分类数据集没有问题，但是SUBJ就会报这个index out of bounds问题。

**Comment by xiami2019** [comment]
问题应该出在evaluation.py的299和301行对delta取索引的位置，这里delta[blen]这个操作感觉没什么实际意义，而SUBJ有些句子的长度减去template后大于delta的第一维长度128，所以索引会出现错误，应该可以直接改成将delta第一维按照对应的batch_size大小裁剪。

**Comment by kongds** [comment]
对于有监督模型transfer的结果，我用的是如下的命令（没有使用delta和mlp）：
```
python evaluation.py \                             
        --model_name_or_path  sup-roberta\                                                          
        --pooler avg\                                                                                
        --mode test\                                                                                  
        --task_set  transfer \                                                                       
        --mask_embedding_sentence \                                                                  
        --mask_embedding_sentence_template "*cls*_This_sentence_:_'_*sent_0*_'_means*mask*.*sep+*"  

python evaluation.py \
        --model_name_or_path  sup-bert\
        --pooler avg\                                   
        --mode test\                                   
        --mask_embedding_sentence \                                                                 
        --task_set  transfer \                                                                      
        --mask_embedding_sentence_template "*cls*_This_sentence_of_\"*sent_0*\"_means*mask*.*sep+*"  
```
另外关于delta在evaluation.py中确实有点问题（没有修改对应的position id），我已经在https://github.com/kongds/Prompt-BERT/commit/093e343ab89c40fc120a1de9dacc6d7ec4e6292e update了

**Comment by xiami2019** [comment]
感谢回复，那请问测试STS任务时需要按照eval_only.sh里那样加上mask_embedding_sentence_delta和mask_embedding_sentence_use_pooler吗

**Comment by kongds** [comment]
是的 在STS任务用eval_only.sh的命令就行了

**Comment by xiami2019** [comment]
这个是为什么呢，按理来说得到sentence embeddings的方式应该都是一样的吧

**Comment by xiami2019** [comment]
感觉evaluation.py可以按照如下的方式修改
```
blen = batch['attention_mask'].sum(-1) - template_len
if delta.shape[0] != pooler_output.shape[0]:
    delta_scale = delta[:pooler_output.shape[0],:]
else:
    delta_scale = delta
if args.mask_embedding_sentence_org_mlp:
    pooler_output -= mlp(delta_scale)
else:
    pooler_output -= delta_scale
```
现在的写法中
```
blen = batch['attention_mask'].sum(-1) - template_len
if args.mask_embedding_sentence_org_mlp:
    pooler_output -= mlp(delta[blen])
else:
    pooler_output -= delta[blen]
```
delta[blen]这个操作应该起到的就是按beln的中的元素取对应delta对应维的向量，但是delta每一维的向量都是相同的，如果blen中出现了超过delta.shape[0]的元素，还是会有问题

**Comment by kongds** [comment]
关于不一样这一点，我觉得delta和mlp是对于bert得到的sentence embeddings再进行处理的一步操作。这个操作对于STS的任务会有帮助，但是对于transfer的task可能会有负面的影响（我觉得可能是mlp和delta这部会损失掉下游分类任务中一部分的信息）。

感谢建议 但是这个写法应该是不正确的。
减去的delta应该按照input_ids的lenght去找到对应的delta 但是pooler_output.shape[0]是padding过后的。

比如input_ids的句子长度分别是10,5,7（blen=torch.LongTensor([10,5,7])） 
那么他们的delta就应该对应delta[10],delta[5],delta[7]

**Comment by xiami2019** [comment]
从get_delta里看，这里的delta[10], delta[5], delta[7]实际上都是相同的tensor，实际上delta矩阵是一个batch_size\*768的矩阵，并且每一个行向量都是相同的，我理解如果要跟序列长度有关的话他应该是一个batch_size\*max_len\*768的三维矩阵。pooler_output的shape[0]应该是batch_size，因为下面一行已经取到了mask token对应的contextual embedding。
```
pooler_output = last_hidden[batch['input_ids'] == tokenizer.mask_token_id]
```
所以感觉是可以这么改的，可以避免blen超出delta.shape[0]造成的问题。

**Comment by kongds** [comment]
delta[10], delta[5], delta[7]不相同吧
他们的position_ids不一样，之前忘记加position_ids了 我在这里更新了https://github.com/kongds/Prompt-BERT/commit/093e343ab89c40fc120a1de9dacc6d7ec4e6292e

**Comment by xiami2019** [comment]
```
def get_delta(model, template, tokenizer, device, args):
    model.eval()

    template = template.replace('*mask*', tokenizer.mask_token)\
                    .replace('*sep+*', '')\
                    .replace('*cls*', '').replace('*sent_0*', ' ')
    # strip for roberta tokenizer
    bs_length = len(tokenizer.encode(template.split(' ')[0].replace('_', ' ').strip())) - 2 + 1
    # replace for roberta tokenizer
    batch = tokenizer([template.replace('_', ' ').strip().replace('   ', ' ')], return_tensors='pt')
    batch['position_ids'] = torch.arange(batch['input_ids'].shape[1]).to(device).unsqueeze(0)
    for k in batch:
        batch[k] = batch[k].repeat(256, 1).to(device)
    batch['position_ids'][:, bs_length:] += torch.arange(256).to(device).unsqueeze(-1)
    m_mask = batch['input_ids'] == tokenizer.mask_token_id

    with torch.no_grad():
        outputs = model(**batch,  output_hidden_states=True, return_dict=True)
        last_hidden = outputs.hidden_states[-1]
        delta = last_hidden[m_mask]
    delta.requires_grad = False
    #import pdb;pdb.set_trace()
    template_len = batch['input_ids'].shape[1]
    return delta, template_len
```
从这个函数里看这个batch中的每一个sample都是一样的template。然后我跑出来也确实是相同的。。。

**Comment by xiami2019** [comment]
他们加的也都是完全一样的position_ids吧，所以应该不会造成encoding的差异？

**Comment by kongds** [comment]
是不一样的吧
<img width="581" alt="image" src="https://user-images.githubusercontent.com/19282154/212457612-256317e9-6697-4982-b4e8-f8417ff5d840.png">
<img width="520" alt="image" src="https://user-images.githubusercontent.com/19282154/212457622-4094ebe5-b793-4bdb-983d-0a026d841c02.png">
这一行修改了position_ids

https://github.com/kongds/Prompt-BERT/blob/093e343ab89c40fc120a1de9dacc6d7ec4e6292e/evaluation.py#L60

**Comment by xiami2019** [comment]
> 是不一样的吧 <img alt="image" width="581" src="https://user-images.githubusercontent.com/19282154/212457612-256317e9-6697-4982-b4e8-f8417ff5d840.png"> <img alt="image" width="520" src="https://user-images.githubusercontent.com/19282154/212457622-4094ebe5-b793-4bdb-983d-0a026d841c02.png"> 这一行修改了position_ids
> 
> https://github.com/kongds/Prompt-BERT/blob/093e343ab89c40fc120a1de9dacc6d7ec4e6292e/evaluation.py#L60

哦哦明白了，感谢您热心回复！
